lib/tests/test_rewrite.rs

```diff
@@ -66,6 +66,127 @@ fn test_restore_tree() {
+    let commit_e = graph_builder.commit_with_parents_and_tree(&[&commit_b], &tree_b);
+    let commit_f =
+        graph_builder.commit_with_parents_and_tree(&[&commit_c, &commit_d, &commit_e], &tree_f);
+    let commit_g = graph_builder.commit_with_parents_and_tree(&[&commit_f], &tree_g);
+    let commit_bd = graph_builder.commit_with_parents_and_tree(&[&commit_a], &tree_d);
+
+    let mut rebaser = DescendantRebaser::new(
+        &settings,
+        tx.mut_repo(),
+        hashmap! {
+            commit_b.id().clone() => hashset!{commit_bd.id().clone()}
+        },
+        hashset! {},
+        RebaseOptions {
+            empty: empty.clone(),
+            ..Default::default()
+        },
+    );
+
+    let new_commit_c = assert_rebased(rebaser.rebase_next().unwrap(), &commit_c, &[&commit_bd]);
+
+    let new_head = match empty {
+        EmptyBehaviour::Keep => {
+            let new_commit_d =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_d, &[&commit_bd]);
+            let new_commit_e =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_e, &[&commit_bd]);
+            let new_commit_f = assert_rebased(
+                rebaser.rebase_next().unwrap(),
+                &commit_f,
+                &[&new_commit_c, &new_commit_d, &new_commit_e],
+            );
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_f])
+        }
+        EmptyBehaviour::AbandonAllEmpty => {
+            assert_abandoned(rebaser.rebase_next().unwrap(), &commit_d, &commit_bd);
+            assert_abandoned(rebaser.rebase_next().unwrap(), &commit_e, &commit_bd);
+            assert_abandoned(rebaser.rebase_next().unwrap(), &commit_f, &new_commit_c);
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_c])
+        }
+        EmptyBehaviour::AbandonNewlyEmpty => {
+            assert_abandoned(rebaser.rebase_next().unwrap(), &commit_d, &commit_bd);
+            let new_commit_e =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_e, &[&commit_bd]);
+            let new_commit_f = assert_rebased(
+                rebaser.rebase_next().unwrap(),
+                &commit_f,
+                &[&new_commit_c, &new_commit_e],
+            );
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_f])
+        }
+    };
+
+    assert!(rebaser.rebase_next().unwrap().is_none());
+    assert_eq!(rebaser.rebased().len(), 5);
+
+    assert_eq!(
+        *tx.mut_repo().view().heads(),
+        hashset! {
+            new_head.id().clone(),
+        }
+    );
+}
+
+#[test]
+fn test_rebase_abandon_all_empty() {
+    test_empty_commit_option(EmptyBehaviour::AbandonAllEmpty)
+}
+
+#[test]
+fn test_rebase_abandon_newly_empty() {
+    test_empty_commit_option(EmptyBehaviour::AbandonNewlyEmpty)
+}
+
+#[test]
+fn test_rebase_keep_empty() {
+    test_empty_commit_option(EmptyBehaviour::Keep)
+}
```

Comment:
> If we have each test case as standing on its own, we can explain the expected behavior in comments before hitting any assertions.

I'm a little confused about this. I can just as easily do this with test cases not standing on their own. I've updated the commit to do so.

I personally really dislike splitting this into each one standing on its own, because these tests are very related to each other, and should share the same test data by design. Generally, I'd agree that tests shouldn't be inlined into a helper if they have any complex logic, but literally the only logic is to match on the enum to decide which assertions to make. This makes it really easy to understand exactly how each of these options should differ.